[Merged by Bors] - feat(algebra/ring/basic): define non-unital, non-associative rings

ocfnash opened the pull request:
This introduces the following typeclasses beneath `semiring`:
  * `non_unital_non_assoc_semiring`
  * `non_unital_semiring`
  * `non_assoc_semiring`

The goal is to use these to support a non-unital, non-associative
algebras.

The typeclass requirements of `subring`, `subsemiring`, and `ring_hom` are relaxed from `semiring` to `non_assoc_semiring`.

Instances of these new typeclasses are added for:
* alias types:
  * `opposite`
  * `ulift`
* convolutive types:
  * `(add_)monoid_algebra`
  * `direct_sum`
  * `set_semiring`
  * `hahn_series`
* elementwise types: 
  * `locally_constant`
  * `pi`
  * `prod`
  * `finsupp`

---
The changes here aim to be the uncontroversial part of https://github.com/leanprover-community/mathlib/pull/6591

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

- [x] depends on #6865
- [x] depends on #7339
- [x] depends on #7346 (`semigroup_with_zero`)
- [x] depends on #7802

ocfnash committed:
feat(algebra/ring/basic): define non-unital, non-associative rings

They key changes here are the introduction of the following:
  * `non_unital_non_assoc_semiring`
  * `non_unital_semiring`
  * `non_assoc_semiring`

The goal is to use these to support a non-unital, non-associative
algebras.

ocfnash commented on src/analysis/special_functions/trigonometric.lean:
```diff
@@ -2744,7 +2744,7 @@ lemma cos_surjective : function.surjective cos :=
   intro x,
   obtain ⟨w, w₀, hw⟩ : ∃ w ≠ 0, 1 * w * w + (-2 * x) * w + 1 = 0,
-  { rcases exists_quadratic_eq_zero one_ne_zero (exists_eq_mul_self _) with ⟨w, hw⟩,
+  { rcases exists_quadratic_eq_zero (@one_ne_zero ℂ _ _) (exists_eq_mul_self _) with ⟨w, hw⟩,
```
For reasons I only partially understand, the new instances added in `algebra/ring/basic` mean we need to give the elaborator a little more help here.

bryangingechen replied:
I'd like to understand this better too, if any more experienced members have any hints.

ocfnash commented on src/data/mv_polynomial/variables.lean:
```diff
@@ -103,7 +103,7 @@ lemma degrees_X' (n : σ) : degrees (X n : mv_polynomial σ R) ≤ {n} :=
 
 @[simp] lemma degrees_X [nontrivial R] (n : σ) : degrees (X n : mv_polynomial σ R) = {n} :=
-(degrees_monomial_eq _ _ one_ne_zero).trans (to_multiset_single _ _)
+(degrees_monomial_eq _ (1 : R) one_ne_zero).trans (to_multiset_single _ _)
```
Similarly, the elaborator needs a bit of help here.

Vierkantor replied:
From `set_option trace.type_context.is_def_eq_detail true`, it looks like the following is happening:

Previously, `one_ne_zero` had the type  `_ ≠ (@has_zero.zero _ (@mul_zero_class.to_has_zero _ (monoid_with_zero.to_mul_zero_class _)))`, exactly what we're looking for. In this PR the goal is instead `_ ≠ @has_zero.zero _ (@mul_zero_class.to_has_zero _ (non_unital_non_assoc_semiring.to_mul_zero_class _)`. So the elaborator starts unfolding a lot of definitions. I think because it unfolds a lot of these definitions, it doesn't get around to unifying this type with the expected type from `trans _ (to_multiset_single _ _)`. So it picks the default instance for `mul_zero_class`, namely `nat`, and continues. Then it needs to check `trans _ (to_multiset_single _ _)`, sees `mv_polynomial` on one side and `nat` on the other, and errors.

Vierkantor replied:
I think the only way to fix this (without touching the elaborator) is to make the `semiring` to `has_zero` diamond go via `monoid_with_zero`. But that is exactly what this PR is supposed to avoid so I think the conclusion is to insert the `(1 : R)` hint.

eric-wieser replied:
The direction taken by the diamond can be adjusted by reordering the base classes + priorities I think, which doesn't go against the goal of the PR

Vierkantor replied:
To be a bit more precise, I think the difference is that we need higher-order unification after this change, and before we only needed first-order unification. And because the higher-order unifier cannot be as powerful as the first-order unifier, it needs more hints.

eric-wieser replied:
Also, that's not to say doing that reordering is necessarily worth the effort.

ocfnash replied:
Thanks @Vierkantor and @eric-wieser for these remarks.

Vierkantor replied:
I'll go ahead and mark this conversation as resolved since we are OK with inserting the hint.

jcommelin commented on src/algebra/group_with_zero/defs.lean:
```diff
@@ -54,7 +54,13 @@ class no_zero_divisors (M₀ : Type*) [has_mul M₀] [has_zero M₀] : Prop :=
+and right absorbing.
+
+TODO: add the missing conversion from `monoid_with_zero` to `semigroup_with_zero`. -/
+@[protect_proj] class semigroup_with_zero (S₀ : Type*) extends semigroup S₀, mul_zero_class S₀.
```
Could you please add an instance from `monoid_with_zero` to `semigroup_with_zero`.

ocfnash replied:
OK sure. I'm holding off on doing this until I've resolved the issues with the new `monoid` instances since changes this deep are so slow to test and have unpredictable impact. Once that is resolved, I'll take this up.

ocfnash replied:
I've just added https://github.com/leanprover-community/mathlib/pull/6786/commits/34790d40dab423e4452868954a3ddb713ef23231 which attempts to do this. After a lot of pain associated with making changes like this recently, I have no idea if this will build successfully or not, but I pray it does.

ocfnash replied:
Success, thank goodness.

eric-wieser commented on src/algebra/group/defs.lean:
```diff
@@ -174,14 +174,32 @@ theorem mul_left_inj (a : G) {b c : G} : b * a = c * a ↔ b = c :=
 @[ancestor add_semigroup has_zero]
 class add_monoid (M : Type u) extends add_semigroup M, has_zero M :=
 (zero_add : ∀ a : M, 0 + a = a) (add_zero : ∀ a : M, a + 0 = a)
+
```
Similarlyto the comment below (https://github.com/leanprover-community/mathlib/pull/6786/files#r598136450), the instances are missing that go from `monoid` to `mul_one_class`. I remember you tried in the last PR by adjusting the `extends` clauses, but ran into lots of problems.

Can you try again, this time adding the instances manually?

ocfnash replied:
I've just added https://github.com/leanprover-community/mathlib/pull/6786/commits/f867c47f6eb458bb9fcfdf30c6941fb14a2a77c5 which is one last attempt to do this using `extends`.

I'll add the instances manually if that effort is unsuccessful (e.g., if I give up in frustration at the painfully-slow debug cycle again).

ocfnash replied:
I can't believe it. After just one more fix, it now builds 🎉

eric-wieser commented on src/algebra/ring/basic.lean:
```diff
@@ -101,6 +116,18 @@ class semiring (α : Type u) extends add_comm_monoid α, monoid_with_zero α, di
+
+@[priority 100] -- see Note [lower instance priority]
+instance semiring.non_assoc_semiring : non_assoc_semiring α :=
+{ ..‹semiring α› }
```
I wonder; does lowering these to `99` solve the elaborator issues you point out below?

ocfnash replied:
I tried setting the priority to `99` on all three instances here (simultaneously) but the elaborator issues remained.

ocfnash committed:
Change `monoid` / `add_monoid` ancestors

I decided to make one last attempt at this. Let's see if it builds.

ocfnash committed:
Fix timeout caused by change to `monoid` / `add_monoid` ancestors

ocfnash committed:
Add `monoid_with_zero.to_semigroup_with_zero`

I _pray_ that this doesn't send me down another rabbit hole of failed builds.

ocfnash committed:
Remove redundant ancestors from `monoid` / `add_monoid`.

eric-wieser commented on src/algebra/group/defs.lean:
```diff
@@ -174,14 +174,30 @@ theorem mul_left_inj (a : G) {b c : G} : b * a = c * a ↔ b = c :=
+class add_monoid (M : Type u) extends add_semigroup M, add_zero_class M
+
 attribute [to_additive] monoid
 
```
Can you weaken the typeclass arguments for `one_mul` and `mul_one` below? Just moving them to a section taking `[mul_one_class M]` should do the job.

ocfnash replied:
Right, good point. I've attempted to do this in https://github.com/leanprover-community/mathlib/pull/6786/commits/dd80907a7897772c0ecf164c5fb6be260b137cd4

In just three short hours we'll know whether it builds successfully!

eric-wieser commented on src/algebra/pointwise.lean:
```diff
@@ -132,8 +132,8 @@ instance [semigroup α] : semigroup (set α) :=
-  one_mul := λ s, by { simp only [← singleton_one, singleton_mul, one_mul, image_id'] },
+{ mul_one := λ s, by { rw [← singleton_one, mul_singleton], simp only [mul_one, image_id'], },
+  one_mul := λ s, by { rw [← singleton_one, singleton_mul], simp only [one_mul, image_id'], },
   ..set.semigroup,
```
Any idea what happened here?

ocfnash replied:
This is a good question and I don't understand it properly.

Ordinarily I'd be concerned but in this case I am content to be a little ignorant since the number of changes that I was forced to make as a result of the change the definitions of `monoid`, `add_monoid` are very small and these are very heavily-used types in a very large codebase.

Some remarks for what they're worth as follows. There is a difference in behaviour between `simp only [← singleton_one]` and `rw ← singleton_one`. The former turns the goal `s * 1 = s` into `semigroup.mul s {1} = s` whereas the latter produces `s * {1} = s`. The next lemma `mul_singleton` will only fire (either via `rw` or via `simp`) on the goal in the form `s * {1} = s` so we have to use `rw ← singleton_one`. There is then a similar story for `mul_singleton` itself so I split that off into the `rw` also. However the final two lemmas (`mul_one` and `image_id'`) actually need to be `simp` lemmas. Utterly gross and I do not understand but frankly I'm absolutely out of desire to spend more time on this.

kim-em replied:
We need `simp` for the `mul_one` because it is underneath a binder where `rw` can't reach.

kim-em replied:
Okay, in the interests of getting this PR moving, I've just added a comment in the sources here explaining the reason for needing both `simp only` and `rw`.

I also wish I knew why `simp only` was introducing the `semigroup.mul`. :-(

kim-em replied:
Ah, I found a better fix, adding `mul := (*)`, following the trick Gabriel used in https://github.com/leanprover-community/mathlib/pull/6786/commits/7a913e262fb8c4453204e84fb382b40e8157b3af. Now everything works by `simp only`.

ocfnash committed:
Weaken typeclass hypotheses for `mul_one` and `one_mul`.

eric-wieser commented on src/algebra/group/defs.lean:
```diff
@@ -174,26 +174,47 @@ theorem mul_left_inj (a : G) {b c : G} : b * a = c * a ↔ b = c :=
+section monoid
+variables {M : Type u} [monoid M]
 
 attribute [ematch] add_zero zero_add -- TODO(Mario): Make to_additive transfer this
```
Since it doesn't show up in the diff view; this should move to line 203, but won't affect the build.

ocfnash replied:
I've just pushed a commit which fixes this.

gebner committed:
Fix type tags.

ocfnash committed:
Move misplaced attribute statement

kim-em committed:
add comment

kim-em committed:
a better fix

kim-em committed:
another similar fix

kim-em committed:
a bit more

kim-em committed:
missing bits of API pointed out by Eric

eric-wieser commented on src/algebra/pointwise.lean:
```diff
@@ -132,7 +132,8 @@ instance [semigroup α] : semigroup (set α) :=
 
 @[to_additive set.add_monoid]
 instance [monoid α] : monoid (set α) :=
```
Ideally we'd have this instance for the new `mul_one_class (set α)` too

kim-em replied:
Done.

kim-em committed:
a little more

kim-em committed:
pi, inj_surj, locally_constant

kim-em committed:
prod.semigroup_with_zero

kim-em committed:
prod.non_*_semiring

eric-wieser commented on src/algebra/ring/basic.lean:
```diff
@@ -91,6 +91,21 @@ protected def function.surjective.distrib {S} [distrib R] [has_add S] [has_mul S
+/-- A unital but not-necessarily-associative semiring. -/
+@[protect_proj, ancestor non_unital_non_assoc_semiring mul_one_class]
+class non_assoc_semiring (α : Type u) extends
+  non_unital_non_assoc_semiring α, mul_one_class α
```
Something else I've realized is missing - we should have a `non_assoc_semiring` instance on a `subsemiring` (and a `subring`) of a `non_assoc_semiring`.

kim-em replied:
Given that those entire files will admit generalizations now, perhaps these instances should wait for another PR.

kim-em replied:
We'd also need to make `subsemigroup`, and maybe even `subhas_mul`...?

eric-wieser replied:
Adding new subobjects is definitely out of scope here, but adding instances of a typeclass on existing subobjects feels in scope (if we want it to be).

kim-em committed:
equiv.non_*_semiring

kim-em committed:
pointwise

kim-em committed:
upgrade strength of constructions in data/finsupp/pointwise

kim-em committed:
line length

eric-wieser committed:
weaken the typeclass restrictions for the basic submonoid operations

eric-wieser committed:
generalize monoid_hom to mul_one_class

eric-wieser committed:
Fix Mon/basic

ocfnash committed:
Add type ascription to fix elaboration failure

ocfnash committed:
Add type ascription to fix elaboration failure

ocfnash committed:
Add type ascriptions to fix elaboration failure

ocfnash committed:
Manual conflicts resolve

eric-wieser commented on src/algebra/group/defs.lean:
```diff
@@ -211,6 +211,21 @@ mul_one_class.mul_one
 
 attribute [ematch] add_zero zero_add -- TODO(Mario): Make to_additive transfer this
 
+end mul_one_class
```
The changes to this file look like a bad merge

ocfnash replied:
Yeah, this was git's work! I've removed them locally and I'm just looking over the rest.

eric-wieser commented on src/data/equiv/transfer_instance.lean:
```diff
@@ -141,11 +149,30 @@ protected def mul_one_class [mul_one_class β] : mul_one_class α :=
 let one := e.has_one, mul := e.has_mul in
 by resetI; apply e.injective.mul_one_class _; intros; exact e.apply_symm_apply _
 
+/-- Transfer `mul_zero_class` across an `equiv` -/
```
I think this hunk can be removed, and is also a merge conflict

ocfnash replied:
It's toast.

ocfnash committed:
Tidy up after bad merge

ocfnash committed:
Tidy up after bad merge'

eric-wieser commented on src/algebra/group/inj_surj.lean:
```diff
@@ -215,6 +225,16 @@ protected def semigroup [semigroup M₁] (f : M₁ → M₂) (hf : surjective f)
+  [semigroup_with_zero M₁] [has_zero M₂] (f : M₁ → M₂) (hf : surjective f)
+  (zero : f 0 = 0) (mul : ∀ x y, f (x * y) = f x * f y) :
+  semigroup_with_zero M₂ :=
+{ zero_mul := hf.forall.2 $ λ x, by erw [← zero, ← mul, zero_mul],
```
You should be able to use `..` on the mul_zero_class here, instead of proving these two fields

ocfnash replied:
Agreed but `function.injective.mul_zero_class` is in `algebra/group_with_zero/basic.lean` which is not imported. I'm reluctant to change imports just for this, especially as it may be that the right solution is to address the division of such lemmas mentioned in my comment [below](https://github.com/leanprover-community/mathlib/pull/6786#discussion_r606857314).

ocfnash replied:
Oh I guess maybe you're saying I should just move the injective / surjective lemmas for `semigroup_with_zero` for this very reason.

ocfnash replied:
OK now done.

eric-wieser commented on src/algebra/group/inj_surj.lean:
```diff
@@ -44,6 +44,16 @@ protected def semigroup [semigroup M₂] (f : M₁ → M₂) (hf : injective f)
+{ zero_mul := λ x, hf $ by erw [mul, zero, zero_mul],
+  mul_zero := λ x, hf $ by erw [mul, zero, mul_zero],
+  ..‹has_zero M₁›, .. hf.semigroup f mul }
+
```
I think this belongs in a different file next to the one for `monoid_with_zero`:

https://github.com/leanprover-community/mathlib/blob/f29b0b49badab1bcbe338cb79d102e36e79dce09/src/algebra/group_with_zero/basic.lean#L183-L213

ocfnash replied:
What is the logic here?

I see that `algebra/group_with_zero/basic.lean` contains these inj/surj lemms for:

- `mul_zero_class`
- `no_zero_divisors`
- `monoid_with_zero`
- `comm_monoid_with_zero`
- `cancel_monoid_with_zero`
- `group_with_zero`
- `group_with_zero_div`
- `comm_group_with_zero`
- `comm_group_with_zero_div`

whereas `algebra/group/inj_surj.lean` contains such lemmas for:

- `semigroup`
- `semigroup_with_zero` <-- the lemma in question here
- `comm_semigroup`
- `left_cancel_semigroup`
- `right_cancel_semigroup`
- `mul_one_class`
- `monoid`
- `left_cancel_monoid`
- `comm_monoid`
- `div_inv_monoid`
- `group`
- `group_div`
- `comm_group`
- `comm_group_div`

ocfnash replied:
Never mind, I think I see the motivation now.

eric-wieser commented on src/algebra/group/prod.lean:
```diff
@@ -84,6 +84,11 @@ instance [semigroup M] [semigroup N] : semigroup (M × N) :=
+instance [semigroup_with_zero M] [semigroup_with_zero N] : semigroup_with_zero (M × N) :=
+{ zero_mul := assume a, prod.rec_on a $ λa b, mk.inj_iff.mpr ⟨zero_mul _, zero_mul _⟩,
+  mul_zero := assume a, prod.rec_on a $ λa b, mk.inj_iff.mpr ⟨mul_zero _, mul_zero _⟩,
+  .. prod.has_zero, .. prod.semigroup }
```
```suggestion
{ .. prod.mul_zero_class, .. prod.semigroup }
```

eric-wieser commented on src/data/equiv/transfer_instance.lean:
```diff
@@ -124,6 +124,14 @@ protected def semigroup [semigroup β] : semigroup α :=
+  mul_zero := by simp [mul_def, zero_def],
+  ..equiv.has_zero e,
+  ..equiv.has_mul e,
+  ..equiv.semigroup e }
```
This no longer matches the style of the rest of the proofs in this file. I guess it should be something like

```suggestion
let mul := e.has_mul, zero := e.has_zero in
by resetI; apply e.injective.semigroup_with_zero _; intros; exact e.apply_symm_apply _
```

eric-wieser commented on src/data/equiv/transfer_instance.lean:
```diff
@@ -144,8 +152,12 @@ by resetI; apply e.injective.mul_one_class _; intros; exact e.apply_symm_apply _
+  ..equiv.has_mul e,
+  ..equiv.has_one e,
+  ..equiv.semigroup e,
+  ..equiv.mul_one_class e, }
```
I think these lines can be left as they were

eric-wieser commented on src/data/equiv/transfer_instance.lean:
```diff
@@ -165,10 +177,31 @@ protected def comm_group [comm_group β] : comm_group α :=
+/-- Transfer `non_assoc_semiring` across an `equiv` -/
+protected def non_assoc_semiring [non_assoc_semiring β] : non_assoc_semiring α :=
+{ ..equiv.mul_one_class e,
+  ..equiv.non_unital_non_assoc_semiring e }
```
As above, lets prove these with `e.injective.<...>` for consistency.

eric-wieser commented on src/data/equiv/transfer_instance.lean:
```diff
@@ -165,10 +177,31 @@ protected def comm_group [comm_group β] : comm_group α :=
-let zero := e.has_zero, add := e.has_add, one := e.has_one, mul := e.has_mul in
-by resetI; apply e.injective.semiring _; intros; exact e.apply_symm_apply _
+{ ..equiv.non_unital_semiring e,
+  ..equiv.non_assoc_semiring e }
```
No need for the change here.

eric-wieser commented on src/data/finsupp/pointwise.lean:
```diff
@@ -48,17 +48,21 @@ begin
 end
 
-instance [semiring β] : semigroup (α →₀ β) :=
+instance [semigroup_with_zero β] : semigroup (α →₀ β) :=
```
Can the conclusion be promoted to a `semigroup_with_zero (α →₀ β)`?

ocfnash replied:
Good point. Now done in https://github.com/leanprover-community/mathlib/pull/6786/commits/6791a868c55c8eff2afa63d81b1cfb775779b3de

ocfnash committed:
Remove redundant type ascriptions.

These were added in https://github.com/leanprover-community/mathlib/pull/6786/commits/274ceb4987e8b19f0b0cc8fd66248bed9db8a427 but are no longer necessary after a merge from master with https://github.com/leanprover-community/mathlib/pull/6865

ocfnash committed:
Update src/algebra/group/prod.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

ocfnash committed:
Update src/data/equiv/transfer_instance.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

ocfnash committed:
Style fix

ocfnash committed:
Use consistent style for transfer defs.

ocfnash committed:
Upgrade `finsupp.semigroup` instance to `finsupp.semigroup_with_zero` instance.

Also updating an out-of-date comment about the convolution product.

ocfnash committed:
Move `function.injective.semigroup_with_zero`, `function.surjective.semigroup_with_zero`

ocfnash committed:
Revert import change

eric-wieser commented on src/algebra/pointwise.lean:
```diff
@@ -136,6 +136,26 @@ instance [semigroup α] : semigroup (set α) :=
+
+instance [semigroup_with_zero α] : semigroup_with_zero (set α) :=
+{ ..set.mul_zero_class,
+  ..set.semigroup }
```
These instances are a bad idea because they put a different `0` on `set α` to what `has_zero` does, and should be restricted to `set_semiring`

eric-wieser replied:
I've reworked this bit of the PR in #7050

eric-wieser committed:
Merge remote-tracking branch 'origin/master' into non_unital_assoc_rings

eric-wieser commented on src/data/finsupp/pointwise.lean:
```diff
@@ -55,17 +52,26 @@ instance : mul_zero_class (α →₀ β) :=
+  mul       := (*),
+  mul_zero  := λ f, by { ext, simp only [mul_apply, zero_apply, mul_zero], },
+  zero_mul  := λ f, by { ext, simp only [mul_apply, zero_apply, zero_mul], }, }
+
```
This one is repeated above as a result of a bad merge
```suggestion
```

eric-wieser committed:
Update src/data/finsupp/pointwise.lean

eric-wieser commented on src/algebra/ring/prod.lean:
```diff
@@ -24,13 +24,28 @@ variables {R : Type*} {R' : Type*} {S : Type*} {S' : Type*} {T : Type*} {T' : Ty
   left_distrib := λ a b c, mk.inj_iff.mpr ⟨left_distrib _ _ _, left_distrib _ _ _⟩,
   right_distrib := λ a b c, mk.inj_iff.mpr ⟨right_distrib _ _ _, right_distrib _ _ _⟩,
-  .. prod.add_comm_monoid, .. prod.monoid }
+  .. prod.add_comm_monoid, .. prod.has_mul }
```
When looking at this I noticed we're missing a `distrib` instance - extracted as #7069

eric-wieser committed:
Merge branch 'master' into non_unital_assoc_rings

eric-wieser committed:
Merge branch 'master' into non_unital_assoc_rings

eric-wieser committed:
Update pointwise.lean

eric-wieser commented on src/algebra/ring/basic.lean:
```diff
@@ -91,32 +91,115 @@ protected def function.surjective.distrib {S} [distrib R] [has_add S] [has_mul S
+/-- A unital but not-necessarily-associative semiring. -/
+@[protect_proj, ancestor non_unital_non_assoc_semiring mul_one_class]
+class non_assoc_semiring (α : Type u) extends
+  non_unital_non_assoc_semiring α, mul_one_class α
```
```suggestion
  non_unital_non_assoc_semiring α, mul_zero_one_class α
```

eric-wieser committed:
Update src/algebra/ring/basic.lean

eric-wieser commented on src/algebra/group_with_zero/defs.lean:
```diff
@@ -57,10 +57,19 @@ export no_zero_divisors (eq_zero_or_eq_zero_of_mul_eq_zero)
-/-- A type `M` is a “monoid with zero” if it is a monoid with zero element, and `0` is left
+/-- A type `S₀` is a "semigroup with zero” if it is a semigroup with zero element, and `0` is left
+and right absorbing. -/
+@[protect_proj] class semigroup_with_zero (S₀ : Type*) extends semigroup S₀, mul_zero_one_class S₀.
```
```suggestion
@[protect_proj] class semigroup_with_zero (S₀ : Type*) extends semigroup S₀, mul_zero_class S₀.
```
Whoops

eric-wieser committed:
Update src/algebra/group_with_zero/defs.lean

ocfnash committed:
Add type annotation to fix build

eric-wieser committed:
Improve instances for monoid_algebra

eric-wieser committed:
add a semigroup_with_zero instance for with_zero

eric-wieser committed:
Add instances for opposites

eric-wieser committed:
fix

eric-wieser committed:
add missing ulift instances

eric-wieser committed:
feat(algebra/group_with_zero): add semigroup_with_zero

Co-authored-by: Oliver Nash <github@olivernash.org>

eric-wieser commented on src/algebra/opposites.lean:
```diff
@@ -138,10 +138,32 @@ instance [distrib α] : distrib (opposite α) :=
+
+instance [mul_zero_one_class α] : mul_zero_one_class (opposite α) :=
+{ .. opposite.mul_zero_class α, .. opposite.mul_one_class α }
+
```
I've extracted these to #7339

eric-wieser committed:
Merge remote-tracking branch 'origin/master' into non_unital_assoc_rings

eric-wieser committed:
reorder definitions

eric-wieser committed:
Merge remote-tracking branch 'origin/eric-wieser/semigroup_with_zero' into non_unital_assoc_rings

eric-wieser committed:
Merge remote-tracking branch 'origin/master' into eric-wieser/semigroup_with_zero

eric-wieser committed:
Merge branch 'eric-wieser/semigroup_with_zero' into non_unital_assoc_rings

eric-wieser commented on src/algebra/pointwise.lean:
```diff
@@ -323,27 +323,28 @@ instance set_semiring.add_comm_monoid : add_comm_monoid (set_semiring α) :=
+{ ..set.mul_one_class,
+  ..set_semiring.non_unital_non_assoc_semiring }
 
 instance set_semiring.semigroup_with_zero [semigroup α] : semigroup_with_zero (set_semiring α) :=
```
```suggestion
```

eric-wieser commented on src/algebra/pointwise.lean:
```diff
@@ -323,27 +323,28 @@ instance set_semiring.add_comm_monoid : add_comm_monoid (set_semiring α) :=
+  ..set_semiring.non_unital_non_assoc_semiring }
 
 instance set_semiring.semigroup_with_zero [semigroup α] : semigroup_with_zero (set_semiring α) :=
 { ..set_semiring.mul_zero_class, ..set.semigroup }
```
```suggestion
```

eric-wieser commented on src/algebra/pointwise.lean:
```diff
@@ -323,27 +323,28 @@ instance set_semiring.add_comm_monoid : add_comm_monoid (set_semiring α) :=
+instance set_semiring.non_assoc_semiring [mul_one_class α] : non_assoc_semiring (set_semiring α) :=
+{ ..set.mul_one_class,
+  ..set_semiring.non_unital_non_assoc_semiring }
 
```
```suggestion
```

eric-wieser committed:
Apply suggestions from code review

eric-wieser committed:
style: reorder to make the overall diff smaller

eric-wieser committed:
Try using extends

eric-wieser committed:
fix

eric-wieser committed:
fix

eric-wieser committed:
Merge remote-tracking branch 'origin/master' into non_unital_assoc_rings

eric-wieser committed:
Restore an ugly annotation

eric-wieser committed:
Restore the other two ugly annotations

eric-wieser committed:
Merge remote-tracking branch 'origin/master' into non_unital_assoc_rings

ocfnash committed:
Merge branch 'master' into non_unital_assoc_rings

eric-wieser committed:
feat(algebra/ring): relax typeclass assumptions for subsemiring, ring_hom, and ring_equiv to non_assoc_semiring

eric-wieser committed:
Merge branch 'master' into non_unital_assoc_rings

eric-wieser committed:
resolve merge issues

eric-wieser committed:
chore(algebra/direct_sum_graded): non_unital_non_assoc_semiring

eric-wieser commented on src/ring_theory/subsemiring.lean:
```diff
@@ -717,36 +726,36 @@ section actions
 /-- Note that this provides `is_scalar_tower S R R` which is needed by `smul_mul_assoc`. -/
 instance
-  [has_scalar α β] [mul_action R α] [mul_action R β] [is_scalar_tower R α β] (S : subsemiring R) :
+  [has_scalar α β] [mul_action R' α] [mul_action R' β] [is_scalar_tower R' α β] (S : subsemiring R') :
```
```suggestion
  [has_scalar α β] [mul_action R' α] [mul_action R' β] [is_scalar_tower R' α β]
  (S : subsemiring R') :
```

eric-wieser committed:
Update src/ring_theory/subsemiring.lean

eric-wieser committed:
Merge remote-tracking branch 'origin/master' into non_unital_assoc_rings

eric-wieser committed:
fix(deprecated/group): generalize some lemmas

eric-wieser committed:
fix(algebra/big_operators/ring): generalize some lemmas

eric-wieser committed:
feat(ring_theory/hahn_series): generalize semiring typeclasses

This isn't a complete generalization of all lemmas, but at least puts the instances in place

eric-wieser committed:
Merge remote-tracking branch 'origin/master' into non_unital_assoc_rings

# Conflicts:
#	src/data/finsupp/pointwise.lean

eric-wieser commented on src/ring_theory/hahn_series.lean:
```diff
@@ -315,6 +315,21 @@ end
+end
+
+end domain
+
```
This change is #7802, which is probably worth merging first just to keep the history manageable

eric-wieser reviewed (approved):
CI needs to run again, but I think I cleaned up all the `distrib` instances that can now be `non_unital_non_assoc_semiring` instances.

eric-wieser committed:
Merge branch 'master' of github.com:leanprover-community/mathlib into non_unital_assoc_rings

eric-wieser commented on src/data/multiset/antidiagonal.lean:
```diff
@@ -77,7 +77,9 @@ begin
   { simp },
   { assume a s ih,
-    simp [ih, add_mul, mul_comm, mul_left_comm, mul_assoc, sum_map_mul_left.symm],
+    have := @sum_map_mul_left α β _,
```
```suggestion
```
